hollow/src/main/java/com/netflix/hollow/core/memory/encoding/HashCodes.java

```diff
@@ -168,13 +168,40 @@ public static int hashInt(int key) {
         return key;
     }
 
-    public static int hashTableSize(int numElements) {
+    /**
+     * Determine size of hash table capable of storing the specified number of elements with a load
+     * factor applied.
+     *
+     * @param numElements number of elements to be stored in the table
+     * @return size of hash table, always a power of 2
+     * @throws IllegalArgumentException when numElements is negative
+     */
+    public static int hashTableSize(int numElements) throws IllegalArgumentException {
+        if (numElements < 0) {
+            throw new IllegalArgumentException("cannot be negative; numElements="+numElements);
+        } else if (numElements > HASH_TABLE_MAX_SIZE) {
+            throw new IllegalArgumentException("exceeds max hash table size; numElements="+numElements);
+        }
+
         if (numElements == 0)
             return 1;
         if (numElements < 3)
             return numElements * 2;
-        int sizeAfterLoadFactor = numElements * 10 / 7;
-        return 1 << (32 - Integer.numberOfLeadingZeros(sizeAfterLoadFactor - 1));
-    }
 
+        // Apply load factor to number of elements and determine next
+        // largest power of 2 that fits in an int
+        int sizeAfterLoadFactor = (int)((long)numElements * 10 / 7);
+        final int size;
+        if (sizeAfterLoadFactor < 0) { // overflow
+            size = HASH_TABLE_MAX_SIZE;
+        } else {
+            int bits = 32 - Integer.numberOfLeadingZeros(sizeAfterLoadFactor - 1);
+            if (bits == 31) {
```

Comment:
Previously, with a large enough `sizeAfterLoadFactor` the `1 << bits` would also overflow. Now we clamp at `HASH_TABLE_MAX_SIZE` (`1 << 30`)